File: tensorflow/compiler/plugin/poplar/kernels/datastream/dataset_benchmark.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include <poplar/DeviceManager.hpp>
#include <poplar/Engine.hpp>
#include <poplar/Tensor.hpp>
#include <poplar/exceptions.hpp>
#include <poplar/replication_factor.hpp>
#include <poputil/TileMapping.hpp>

#include "absl/container/flat_hash_set.h"
#include "include/json/json.h"
#include "tensorflow/compiler/plugin/poplar/driver/poplar_executor.h"
#include "tensorflow/compiler/plugin/poplar/driver/poplar_feed_config.pb.h"
#include "tensorflow/compiler/plugin/poplar/driver/poplar_platform.h"
#include "tensorflow/compiler/plugin/poplar/driver/tools/infeed_allocator.h"
#include "tensorflow/compiler/plugin/poplar/driver/tools/infeed_iterator.h"
#include "tensorflow/compiler/plugin/poplar/driver/tools/io_thread.h"
#include "tensorflow/compiler/plugin/poplar/driver/xla_ipu_common.h"
#include "tensorflow/compiler/plugin/poplar/kernels/custom_kernels_util.h"
#include "tensorflow/compiler/plugin/poplar/kernels/ipu_kernels_common.h"
#include "tensorflow/compiler/tf2xla/shape_util.h"
#include "tensorflow/compiler/tf2xla/type_util.h"
#include "tensorflow/compiler/tf2xla/xla_helpers.h"
#include "tensorflow/compiler/tf2xla/xla_op_kernel.h"
#include "tensorflow/compiler/xla/literal_util.h"
#include "tensorflow/compiler/xla/shape_util.h"
#include "tensorflow/core/common_runtime/dma_helper.h"
#include "tensorflow/core/framework/dataset.h"
#include "tensorflow/core/framework/node_def_util.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/types.h"
#include "tensorflow/core/kernels/data/dataset_utils.h"
#include "tensorflow/core/platform/env.h"
#include "tensorflow/core/platform/mem.h"
#include "tensorflow/core/platform/notification.h"
#include "tensorflow/core/platform/types.h"
#include "tensorflow/core/util/batch_util.h"
#include "tensorflow/core/util/stream_executor_util.h"

namespace tensorflow {
namespace {
void XlaShapesFromAttr(OpKernelConstruction* ctx,
                       std::vector<xla::Shape>& result) {
  std::vector<TensorShape> shapes;
  std::vector<tensorflow::DataType> types;
  OP_REQUIRES_OK(ctx, ctx->GetAttr("output_shapes", &shapes));
  OP_REQUIRES_OK(ctx, ctx->GetAttr("output_types", &types));

  for (unsigned i = 0; i < shapes.size(); ++i) {
    xla::PrimitiveType xla_type;
    OP_REQUIRES_OK(ctx, DataTypeToPrimitiveType(types[i], &xla_type));
    result.emplace_back(TensorShapeToXLAShape(xla_type, shapes[i]));
  }
}

xla::poplarplugin::IOFunction ConsumerThread(
    bool print_stats, bool do_memcpy, uint64 number_of_epochs,
    uint64 elements_per_epochs, Notification* finished_notifiation,
    Json::Value* stats_json, xla::poplarplugin::InfeedIterator* itr) {
  return [=](std::atomic<bool>& cancelled) {
    Json::Value epochs(Json::arrayValue);

    // We only ever have a single replica.
    const size_t replica_id = 0;
    auto queues = itr->GetInfeedQueues();
    auto shapes = itr->GetShapes();

    // Allocate the buffers we copy into.
    std::vector<void*> buffers(shapes.size());
    std::vector<size_t> buffer_sizes(shapes.size());
    std::size_t total_bytes = 0;
    for (uint64 i = 0; i != shapes.size(); ++i) {
      buffer_sizes[i] = xla::ShapeUtil::ByteSizeOf(shapes[i]);
      total_bytes += buffer_sizes[i];
      buffers[i] = port::AlignedMalloc(buffer_sizes[i], 4096);
    }

    // Run for the amount of time user has asked us to.
    for (uint64 i = 0; i != number_of_epochs; ++i) {
      using seconds = std::chrono::duration<float>;
      auto t0 = std::chrono::high_resolution_clock::now();
      for (uint64 j = 0; j != elements_per_epochs; ++j) {
        auto& replica_queues = queues[replica_id];
        for (uint64 k = 0; k != replica_queues.size(); ++k) {
          tensorflow::TensorBuffer* buf;
          // Continue to try and get a buffer unless we have been cancelled.
          while (!replica_queues[k]->TryPop(buf)) {
            if (cancelled) {
              return tensorflow::errors::Aborted("Consumer thread cancelled");
            }
          }
          if (do_memcpy) {
            std::memcpy(buffers[k], buf->data(), buffer_sizes[k]);
          }
          replica_queues[k]->AdvanceReadPosition();
        }
      }
      auto t1 = std::chrono::high_resolution_clock::now();
      Json::Value epoch_stats;
      const float elements_processed = static_cast<float>(elements_per_epochs);
      const float total_bytes_processed = elements_processed * total_bytes;
      const float time_elapsed = static_cast<float>(seconds(t1 - t0).count());
      const float elements_per_second = elements_processed / time_elapsed;
      const float bandwidth =
          total_bytes_processed / (1000000000.f * time_elapsed);

      epoch_stats["elements_processed"] = elements_processed;
      epoch_stats["total_bytes_processed"] = total_bytes_processed;
      epoch_stats["time_elapsed"] = time_elapsed;
      epoch_stats["elements_per_second"] = elements_per_second;
      epoch_stats["bandwidth"] = bandwidth;
      epochs.append(epoch_stats);

      if (print_stats) {
        LOG(INFO) << "Processed: " << elements_per_second
                  << " elements/second.";
        LOG(INFO) << "Bandwidth: " << bandwidth << " GB/s.";
      }
      LOG(INFO) << "Dataset iterator completed epoch " << i << ".";
    }
    // Store stats into the json.
    (*stats_json)["epochs"] = epochs;
    // Notify we have finished.
    finished_notifiation->Notify();
    return Status::OK();
  };
}

xla::poplarplugin::IOFunction ProducerThread(
    xla::poplarplugin::InfeedIterator* itr) {
  return [=](std::atomic<bool>& cancelled) {
    // We only ever have a single replica.
    const size_t replica_id = 0;
    auto queues = itr->GetInfeedQueues();

    while (!cancelled) {
      if (queues[0][0]->IsFull()) {
        VLOG(1) << "Infeed queue is full.";
        continue;
      }

      if (queues[0][0]->IsEmpty()) {
        VLOG(1) << "Infeed queue is empty.";
      }

      std::vector<tensorflow::Tensor> outputs;
      bool end_of_sequence = false;
      TF_RETURN_IF_ERROR(itr->GetNext(&outputs, &end_of_sequence));

      if (end_of_sequence) {
        return tensorflow::errors::OutOfRange(
            "The dataset iterator has reached the end of the dataset.");
      }

      for (size_t j = 0; j < outputs.size(); ++j) {
        auto& queue = queues[replica_id][j];
        TensorBuffer* tb = tensorflow::DMAHelper::buffer(&outputs[j]);
        tb->Ref();
        queue->BlockPush(tb);
        queue->AdvanceWritePosition();
      }
    }
    return Status::OK();
  };
}

}  // namespace

class DatasetBenchmark : public OpKernel {
 public:
  explicit DatasetBenchmark(OpKernelConstruction* ctx) : OpKernel(ctx) {
    OP_REQUIRES_OK(ctx, ctx->GetAttr("print_stats", &print_stats_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("do_memcpy", &do_memcpy_));
    OP_REQUIRES_OK(ctx, ctx->GetAttr("number_of_epochs", &number_of_epochs_));
    OP_REQUIRES_OK(ctx,
                   ctx->GetAttr("elements_per_epochs", &elements_per_epochs_));
    XlaShapesFromAttr(ctx, shapes_);
  }

  ~DatasetBenchmark() override {}

  void Compute(OpKernelContext* ctx) override {
    // Get the flr and create base parameters.
    FunctionLibraryRuntime* flr = ctx->function_library();
    data::IteratorContext::Params params(ctx);

    // Get the dataset
    data::DatasetBase* dataset;
    OP_REQUIRES_OK(ctx, GetDatasetFromVariantTensor(ctx->input(0), &dataset));

    xla::poplarplugin::InfeedAllocator infeed_allocator;
    xla::poplarplugin::InfeedIterator infeed_iterator(
        flr, params, dataset, &infeed_allocator, shapes_, "benchmark");

    infeed_iterator.SetReplicationFactor(1);

    Json::Value stats_json;
    {
      Notification consumer_finished_notifiation;

      // Start the consumer thread.
      xla::poplarplugin::IOThread consumer_thread(
          "consumer",
          ConsumerThread(print_stats_, do_memcpy_, number_of_epochs_,
                         elements_per_epochs_, &consumer_finished_notifiation,
                         &stats_json, &infeed_iterator));

      // Start the producer thread.
      xla::poplarplugin::IOThread producer_thread(
          "producer", ProducerThread(&infeed_iterator));

      // Wait until the consumer thread has finished - note that the thread will
      // be cancelled and destroyed as we exit the scope.
      consumer_finished_notifiation.WaitForNotification();
    }
    // Copy the output out as a string.
    Json::StreamWriterBuilder builder;
    std::stringstream ss;
    std::unique_ptr<Json::StreamWriter> writer(builder.newStreamWriter());
    writer->write(stats_json, &ss);

    Tensor* output_tensor = nullptr;
    OP_REQUIRES_OK(
        ctx, ctx->allocate_output("out", TensorShape({1}), &output_tensor));
    output_tensor->flat<tstring>()(0) = ss.str();
  }

 private:
  bool print_stats_;
  bool do_memcpy_;
  int number_of_epochs_;
  int elements_per_epochs_;
  std::vector<xla::Shape> shapes_;

  TF_DISALLOW_COPY_AND_ASSIGN(DatasetBenchmark);
};  // namespace tensorflow

REGISTER_KERNEL_BUILDER(Name("DatasetBenchmark").Device(DEVICE_CPU),
                        DatasetBenchmark);
}  // namespace tensorflow

File: tensorflow/python/ipu/dataset_benchmark.py
# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# =============================================================================
"""
Dataset benchmarking
~~~~~~~~~~~~~~~~~~~~
"""

from tensorflow.compiler.plugin.poplar.ops import gen_dataset_benchmark
from tensorflow.python.ipu import ipu_infeed_queue
from tensorflow.python.data.ops import dataset_ops


def dataset_benchmark(dataset,
                      number_of_epochs,
                      elements_per_epochs,
                      print_stats=True,
                      apply_debug_options=True,
                      do_memcpy=True):
  """Allows the user to benchmark performance of a `tf.data.Dataset`.

    Args:
      dataset: An instance of `tf.data.Dataset` which will be benchmarked.
      number_of_epochs: The number of epochs this dataset will be run for.
      elements_per_epochs: The number of elements there are in each epoch.
      print_stats: Whether to print statistics about the performance to the
        console.
      apply_debug_options: Whether to apply debug options.
      do_memcpy: Whether to perform a `memcpy` operation which simulates a
        dataset buffer being copied to a Poplar managed buffer.

    Returns:
      A JSON string with performance statistics, which records the following
      metrics every epoch:
        * `elements_processed` - number of elements processed.
        * `total_bytes_processed` - total number of bytes which was processed.
        * `time_elapsed` - the time it took (in seconds) for the epoch to
          complete.
        * `elements_per_second` - number of elements processed per second.
        * `bandwidth` - the bandwidth achieved, measured in GB/s.

    The JSON string returned can be parsed into a native Python JSON library
    (see https://docs.python.org/3/library/json.html).

    Raises:
      TypeError: if `dataset` is not an instance of `tf.data.Dataset`.
      ValueError: if `number_of_epochs` or `elements_per_epochs` is less than 1.
    """
  if number_of_epochs < 1:
    return ValueError("Expected `number_of_epochs` to be at least 1.")
  if elements_per_epochs < 1:
    return ValueError("Expected `elements_per_epochs` to be at least 1.")

  if not isinstance(dataset, dataset_ops.DatasetV2):
    return TypeError("Expected `dataset` argument to be of type "
                     "`tf.data.Dataset`, but got %s "
                     "instead." % (str(dataset)))

  if apply_debug_options:
    dataset = dataset._apply_debug_options()  # pylint: disable=protected-access

  try:
    dataset_variant = dataset._variant_tensor  # pylint: disable=protected-access
  except TypeError:
    dataset_variant = dataset._as_variant_tensor  # pylint: disable=protected-access

  return gen_dataset_benchmark.dataset_benchmark(dataset_variant, print_stats,
                                                 do_memcpy, number_of_epochs,
                                                 elements_per_epochs,
                                                 **dataset._flat_structure)  # pylint: disable=protected-access


def infeed_benchmark(infeed_queue,
                     number_of_epochs,
                     elements_per_epochs,
                     print_stats=True,
                     do_memcpy=True):
  """Allows the user to benchmark performance of an
    `ipu.ipu_infeed_queue.IPUInfeedQueue`.

    Args:
      infeed_queue: An instance of `ipu.ipu_infeed_queue.IPUInfeedQueue` which
        will be benchmarked.
      number_of_epochs: The number of epochs this infeed queue will be run for.
      elements_per_epochs: The number of elements there are in each epoch.
      print_stats: Whether to print statistics about the performance to the
        console.
      do_memcpy: Whether to perform a `memcpy` operation which simulates a
        dataset buffer being copied to a Poplar managed buffer.

    Returns:
      A JSON string with performance statistics, which records the following
      metrics every epoch:
        * `elements_processed` - number of elements processed.
        * `total_bytes_processed` - total number of bytes which was processed.
        * `time_elapsed` - the time it took (in seconds) for the epoch to
          complete.
        * `elements_per_second` - number of elements processed per second.
        * `bandwidth` - the bandwidth achieved, measured in GB/s.

    The JSON string returned can be parsed into a native Python JSON library
    (see https://docs.python.org/3/library/json.html).

    Raises:
      TypeError: if `infeed_queue` is not an instance of
        `ipu.ipu_infeed_queue.IPUInfeedQueue`.
      ValueError: if `number_of_epochs` or `elements_per_epochs` is less than 1.
    """
  if not isinstance(infeed_queue, ipu_infeed_queue.IPUInfeedQueue):
    return TypeError("Expected `infeed_queue` argument to be of type "
                     "`ipu.ipu_infeed_queue.IPUInfeedQueue`, but got %s "
                     "instead." % (str(infeed_queue)))
  # Don't need to apply options because the infeed queue already applies them.
  apply_debug_options = False
  return dataset_benchmark(
      infeed_queue._dataset,  # pylint: disable=protected-access
      number_of_epochs,
      elements_per_epochs,
      print_stats=print_stats,
      apply_debug_options=apply_debug_options,
      do_memcpy=do_memcpy)
